formula-android/src/main/java/com/instacart/formula/fragment/FragmentFlowStore.kt

```diff
@@ -55,37 +56,67 @@ class FragmentFlowStore(
         }
     }
 
-    fun state(environment: FragmentEnvironment): Observable<FragmentFlowState> {
-        val contractShown = visibleContractEvents.map { contract ->
-            { list: List<FragmentContract<*>> ->
-                if (!list.contains(contract)) {
-                    list + contract
-                } else {
-                    // TODO: should we log this duplicate visibility event?
-                    list
-                }
-            }
-        }
+    override fun initialState(input: Unit): FragmentFlowState = FragmentFlowState()
 
-        val contractHidden = hiddenContractEvents.map { contract ->
-            { list: List<FragmentContract<*>> ->
-                list - contract
+    override fun evaluate(
+        input: Unit,
+        state: FragmentFlowState,
+        context: FormulaContext<FragmentFlowState>
+    ): Evaluation<FragmentFlowState> {
+        val rootInput = Binding.Input(
+            component = Unit,
+            activeKeys = state.activeKeys,
+            onStateChanged = context.eventCallback {
+                transition(state.copy(states = state.states.plus(it.key to it)))
             }
-        }
+        )
+        root.bind(context, rootInput)
 
-        val visibleContracts = Observable.merge(contractShown, contractHidden)
-            .scan(emptyList<FragmentContract<*>>()) { list, reducer -> reducer(list) }
-            .distinctUntilChanged()
+        return Evaluation(
+            output = state,
+            updates = context.updates {
+                events(lifecycleEvents) { event ->
+                    val key = event.key
+                    when (event) {
+                        is LifecycleEvent.Removed -> {
+                            val updated = state.copy(
+                                activeKeys = state.activeKeys.minus(key),
+                                states = state.states.minus(key)
+                            )
+                            transition(updated)
+                        }
+                        is LifecycleEvent.Added -> {
+                            // We want to emit an empty state update if key is not handled.
+                            val notHandled = if (!root.binds(key)) {
+                                listOf(Pair(key, KeyState(key, "missing-registration")))
+                            } else {
+                                emptyList()
+                            }
+                            transition(state.copy(
+                                activeKeys = state.activeKeys.plus(key),
+                                states = state.states.plus(notHandled)
+                            ))
+                        }
+                    }
+                }
+
+                events(visibleContractEvents) {
+                    if (state.visibleKeys.contains(it)) {
+                        // TODO: should we log this duplicate visibility event?
```

Comment:
Still not sure about this?

Reply:
Yeah, I copied this over from the other implementation. I think once I add internal logging I will tackle this as well.